src/order/grade.lean

```diff
@@ -0,0 +1,287 @@
+* `grade_bounded_order`: Graded order where minimal elements have minimal grades and maximal
+  elements have maximal grades.
+* `grade`: The grade of an element. Because an order can admit several gradings, the first argument
+  is the order we grade by.
+* `grade_max_order`: Graded orders with maximal elements. All maximal elements have the same grade.
+* `max_grade`: The maximum grade in a `grade_max_order`.
+* `order_embedding.grade`: The grade of an element in a linear order as an order embedding.
+
+## How to grade your order
+
+Here are the translations between common references and our `grade_order`:
+* [Stanley][stanley2012] defines a graded order of rank `n` as an order where all maximal chains
+  have "length" `n` (so the number of elements of a chain is `n + 1`). This corresponds to
+  `grade_bounded_order (fin (n + 1)) α`.
+* [Engel][engel1997]'s ranked orders are somewhere between `grade_order ℕ α` and
+  `grade_min_order ℕ α`, in that he requires `∃ a, is_min a ∧ grade ℕ a + 0` rather than
+  `∀ a, is_min a → grade ℕ a = 0`. He defines a graded order as an order where all minimal elements
+  have grade `0` and all maximal elements have the same grade. This is roughly a less bundled
+  version of `grade_bounded_order (fin n) α`, assuming we discard orders with infinite chains.
+
+## Implementation notes
+
+One possible definition of graded orders is as the bounded orders whose flags (maximal chains)
+all have the same finite length (see Stanley p. 99). However, this means that all graded orders must
+have minimal and maximal elements and that the grade is not data.
+
+Instead, we define graded orders by their grade function, without talking about flags yet.
+
+## References
+
+* [Konrad Engel, *Sperner Theory*][engel1997]
+* [Richard Stanley, *Enumerative Combinatorics*][stanley2012]
+-/
+
+set_option old_structure_cmd true
+
+open finset nat order_dual
+
+variables {𝕆 ℙ α β : Type*}
+
+/-- An `𝕆`-graded order is an order `α` equipped with a strictly monotone function `grade 𝕆 : α → 𝕆`
+which preserves order covering (`covby`). -/
+class grade_order (𝕆 α : Type*) [preorder 𝕆] [preorder α] :=
+(grade : α → 𝕆)
+(grade_strict_mono : strict_mono grade)
+(covby_grade ⦃a b : α⦄ : a ⋖ b → grade a ⋖ grade b)
+
+/-- A `𝕆`-graded order where minimal elements have minimal grades. -/
+class grade_min_order (𝕆 α : Type*) [preorder 𝕆] [preorder α] extends grade_order 𝕆 α :=
+(is_min_grade ⦃a : α⦄ : is_min a → is_min (grade a))
+
+/-- A `𝕆`-graded order where maximal elements have maximal grades. -/
+class grade_max_order (𝕆 α : Type*) [preorder 𝕆] [preorder α] extends grade_order 𝕆 α :=
+(is_max_grade ⦃a : α⦄ : is_max a → is_max (grade a))
+
+/-- A `𝕆`-graded order where minimal elements have minimal grades and maximal elements have maximal
+grades. -/
+class grade_bounded_order (𝕆 α : Type*) [preorder 𝕆] [preorder α]
+  extends grade_min_order 𝕆 α, grade_max_order 𝕆 α
+
+section preorder
+variables [preorder 𝕆]
+
+section preorder
```

Comment:
I'd suggest choosing contrasting section names instead of "preorder" and "preorder".

Other than that, looks good to me

bors d=YaelDillies

Reply:
I prefer to keep names matching the instances, but I differentiated them using comments